File: payload/skel/.config/quickshell/JS/weather.js
.pragma library

function fetchLocationAndWeather(callback) {
    var locXhr = new XMLHttpRequest();
    locXhr.timeout = 5000; // 强制 5 秒超时设置

    locXhr.onreadystatechange = function() {
        if (locXhr.readyState === XMLHttpRequest.DONE) {
            if (locXhr.status === 200) {
                try {
                    var locData = JSON.parse(locXhr.responseText);
                    if (locData.success === false) {
                        callback(null); // API 报错，返回 null
                        return;
                    }

                    var lat = locData.latitude;
                    var lon = locData.longitude;
                    var cityStr = locData.city;
                    if (!cityStr || cityStr.trim() === "") cityStr = locData.region;
                    if (!cityStr || cityStr.trim() === "") cityStr = locData.country;
                    if (!cityStr || cityStr.trim() === "") cityStr = "UNKNOWN";

                    fetchWeatherAPI(lat, lon, cityStr.toUpperCase(), callback);
                } catch(e) {
                    console.log("Location Parse Error:", e);
                    callback(null); // 解析失败，返回 null
                }
            } else {
                console.log("Location Network Error:", locXhr.status);
                callback(null); // 状态码非 200，返回 null
            }
        }
    }
    
    // 捕获底层网络断开和超时
    locXhr.onerror = function() { console.log("Location XHR Error"); callback(null); }
    locXhr.ontimeout = function() { console.log("Location XHR Timeout"); callback(null); }

    locXhr.open("GET", "https://ipwho.is/?t=" + new Date().getTime(), true);
    locXhr.send();
}

function fetchWeatherAPI(lat, lon, city, callback) {
    var url = "https://api.open-meteo.com/v1/forecast?latitude=" + lat + "&longitude=" + lon + 
              "&current=temperature_2m,relative_humidity_2m,apparent_temperature,is_day,precipitation,weather_code,wind_speed_10m,surface_pressure" + 
              "&hourly=temperature_2m,weather_code" + 
              "&daily=weather_code,temperature_2m_max,temperature_2m_min&timezone=auto";
              
    var xhr = new XMLHttpRequest();
    xhr.timeout = 5000; // 天气请求同样增加 5 秒超时

    xhr.onreadystatechange = function() {
        if (xhr.readyState === XMLHttpRequest.DONE) {
            if (xhr.status === 200) {
                var data = JSON.parse(xhr.responseText);
                data.locName = city;
                data.lat = lat;
                data.lon = lon;
                callback(data);
            } else {
                console.log("Weather API Network Error:", xhr.status);
                callback(null);
            }
        }
    }
    
    // 捕获天气接口的网络异常
    xhr.onerror = function() { console.log("Weather XHR Error"); callback(null); }
    xhr.ontimeout = function() { console.log("Weather XHR Timeout"); callback(null); }

    xhr.open("GET", url, true);
    xhr.send();
}

function getMaterialIcon(code) {
    if (code === 0) return "sunny";
    if (code === 1 || code === 2) return "partly_cloudy_day";
    if (code === 3) return "cloudy";
    if (code === 45 || code === 48) return "foggy";
    if (code >= 51 && code <= 67) return "rainy";
    if (code >= 71 && code <= 82) return "snowing";
    if (code >= 95) return "thunderstorm";
    return "cloud";
}

function getWeatherDesc(code) {
    var mapping = {
        0: "Clear", 1: "Mainly Clear", 2: "Partly Cloudy", 3: "Overcast",
        45: "Fog", 48: "Rime Fog", 51: "Drizzle", 61: "Rain", 71: "Snow", 95: "Storm"
    };
    return mapping[code] || "Cloudy";
}

File: aur/my-niri-desk/payload/skel/.config/quickshell/scripts/weather.py
#!/usr/bin/env python3
import json
import time
import os
import urllib.request
import sys
import ssl
import datetime

# ================= 配置区域 =================
CACHE_FILE = "/tmp/qs_weather_cache.json"
CACHE_DURATION = 1800
ssl._create_default_https_context = ssl._create_unverified_context

WEATHER_CODES = {
    0: "Clear",
    1: "Mainly Clear",
    2: "Partly Cloudy",
    3: "Overcast",
    45: "Fog",
    48: "Rime Fog",
    51: "Drizzle",
    53: "Drizzle",
    55: "Drizzle",
    61: "Rain",
    63: "Rain",
    65: "Heavy Rain",
    71: "Snow",
    73: "Snow",
    75: "Heavy Snow",
    80: "Showers",
    81: "Showers",
    82: "Violent Showers",
    95: "Thunderstorm",
    96: "Thunderstorm",
    99: "Thunderstorm",
}


def get_weather_desc(code):
    return WEATHER_CODES.get(code, "Unknown")


def get_current_location():
    try:
        with urllib.request.urlopen("https://ipapi.co/json/", timeout=3) as response:
            content = response.read().decode("utf-8")
            if not content:
                return None, None, None, False
            data = json.loads(content)
            if not isinstance(data, dict):
                return None, None, None, False
            lat, lon, city = (
                data.get("latitude"),
                data.get("longitude"),
                data.get("city", "Unknown"),
            )
            if lat and lon:
                return lat, lon, city, True
    except Exception:
        pass
    return None, None, None, False


def load_cache():
    if os.path.exists(CACHE_FILE):
        try:
            with open(CACHE_FILE, "r") as f:
                data = json.load(f)
                if isinstance(data, dict):
                    return data
        except:
            pass
    return None


def save_cache(data):
    try:
        with open(CACHE_FILE, "w") as f:
            f.write(json.dumps(data))
    except:
        pass


def fetch_open_meteo(lat, lon, city):
    # 【新增】追加 daily 参数，请求未来 7 天的最高温度和天气代码，并自动适应当地时区
    url = f"https://api.open-meteo.com/v1/forecast?latitude={lat}&longitude={lon}&current_weather=true&daily=weathercode,temperature_2m_max&timezone=auto"
    req = urllib.request.Request(url, headers={"User-Agent": "Quickshell-Widget"})

    with urllib.request.urlopen(req, timeout=5) as response:
        content = response.read().decode("utf-8")
        if not content:
            raise Exception("Empty Response")
        raw = json.loads(content)
        if not isinstance(raw, dict) or "current_weather" not in raw:
            raise Exception("Invalid API Response")

        current = raw["current_weather"]
        is_day_bool = True if current.get("is_day", 1) == 1 else False

        # 【新增】解析未来一周数据
        daily = raw.get("daily", {})
        d_times = daily.get("time", [])
        d_codes = daily.get("weathercode", [])
        d_maxs = daily.get("temperature_2m_max", [])

        forecast_list = []
        # 提取明天开始的连续 6 天预测
        for i in range(1, min(7, len(d_times))):
            try:
                dt = datetime.datetime.strptime(d_times[i], "%Y-%m-%d")
                day_name = dt.strftime("%a")  # 格式化为 Mon, Tue 等
                forecast_list.append(
                    {
                        "day": day_name,
                        "temp": f"{round(d_maxs[i])}°",
                        "desc": get_weather_desc(d_codes[i]),
                    }
                )
            except Exception:
                pass

        return {
            "temp": f"{current['temperature']}°",
            "desc": get_weather_desc(current["weathercode"]),
            "city": city,
            "isDay": is_day_bool,
            "forecast": forecast_list,  # 注入预测数组
            "timestamp": time.time(),
        }


def main():
    cur_lat, cur_lon, cur_city, loc_success = get_current_location()
    cache = load_cache()
    has_valid_cache = isinstance(cache, dict)
    use_cache = False

    if has_valid_cache:
        is_fresh = (time.time() - cache.get("timestamp", 0)) < CACHE_DURATION
        is_same_location = (
            str(cache.get("city")) == str(cur_city) if loc_success else True
        )
        if (loc_success and is_same_location and is_fresh) or not loc_success:
            use_cache = True

    if use_cache and has_valid_cache:
        print(json.dumps(cache))
    else:
        try:
            if not loc_success:
                raise Exception("Loc Failed")
            weather_data = fetch_open_meteo(cur_lat, cur_lon, cur_city)
            save_cache(weather_data)
            print(json.dumps(weather_data))
        except Exception:
            if has_valid_cache:
                print(json.dumps(cache))
            else:
                print(
                    json.dumps(
                        {
                            "temp": "--",
                            "desc": "Offline",
                            "city": "Error",
                            "isDay": True,
                            "forecast": [
                                {"day": "N/A", "temp": "--", "desc": "Unknown"}
                            ]
                            * 6,
                        }
                    )
                )


if __name__ == "__main__":
    main()
